### PR TITLE
OSDOCS-2681: MetalLB and BGP

### PR DESCRIPTION
**Pull request by mikemckiernan**
Fixes https://issues.redhat.com/browse/OSDOCS-2681

----

* Add BGP peer custom resource.
* Add two `bgp` protocol address pool examples.
  One is simple and one shows custom advertisements.

----

Preview starts at the [About MetalLB and the MetalLB Operator](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/about-metallb.html) page:
* Descs for new custom resources: `BGPPeer` and `BFDProfile`.
* Concepts for BGP mode (incl draft topology diagram).
* Limitations for BGP mode.

The [Configuring address pools](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/metallb-configure-address-pools.html) page includes:
* Field descs for the BGP advertisements.
* Two new examples for BGP, starting at [a simple IPv4](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/metallb-configure-address-pools.html#example-ipv4-address-pool-and-bgp-mode) example.

[Configuring BGP peers](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/metallb-configure-bgp-peers.html) is new.

[Configuring BFD profiles](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/metallb-configure-bfd-profiles.html) is new.

[MetalLB troubleshooting and support](https://deploy-preview-38580--osdocs.netlify.app/openshift-enterprise/latest/networking/metallb/metallb-troubleshoot-support.html) is new.

### CODE REVIEW

File: modules/nw-metallb-example-addresspool.adoc [hunk 1]
```diff
@@ -59,3 +62,66 @@ spec:
...
+
+For BGP mode, you must set the `protocol` field set to `bgp`.
+Other address pool custom resource fields, such as `autoAssign`, also apply to BGP mode.
+
+In the following example, the peer BGP routers receive one `172.18.0.x/32` route for each load-balancer IP address that MetalLB assigns to a service.
+Because the `localPref` and `communities` fields are not specified, the routes are advertised with `localPref` set to zero and no BGP communities.
+
+[source,yaml]
+----
+apiVersion: metallb.io/v1alpha1
```
Comment (fedepaol) [inline modules/nw-metallb-example-addresspool.adoc:76]: this is moving to beta.

File: modules/nw-metallb-example-addresspool.adoc [hunk 2]
```diff
@@ -59,3 +62,66 @@ spec:
...
+    - 172.18.0.100-172.18.0.255
+----
+
+== Example: BGP mode with custom advertisement
+
+You can specify sophisticated custom advertisements.
+
+[source,yaml]
+----
+apiVersion: metallb.io/v1alpha1
```
Comment (fedepaol) [inline modules/nw-metallb-example-addresspool.adoc:93]: same

File: networking/metallb/metallb-configure-bgp-peers.adoc [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+// Examples
+== Example BGP peer configurations
+
+=== Example: Limit which nodes connect to a BGP peer
+
+You can specify the node selectors field to control which nodes can connect to a BGP peer.
+
+[source,yaml]
+----
+apiVersion: metallb.io/v1alpha1
```
Comment (fedepaol) [inline networking/metallb/metallb-configure-bgp-peers.adoc:135]: nit: beta

File: networking/metallb/metallb-configure-bgp-peers.adoc [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+----
+
+=== Example: Specify a BFD profile for a BGP peer
+
+You can specify a BFD profile to associate with BGP peers.
+BFD compliments BGP by providing more rapid detection of communication failures between peers than BGP alone.
+
+[source,yaml]
+----
+apiVersion: metallb.io/v1alpha1
```
Comment (fedepaol) [inline networking/metallb/metallb-configure-bgp-peers.adoc:158]: same

**Comment by fedepaol** [comment]
I think we need a section about deprecating addresspool v1alpha1 too.

File: networking/metallb/metallb-configure-bgp-peers.adoc [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+|`integer`
+|Optional: Specifies the duration, in seconds, for the hold time to propose to the BGP peer.
+The minimum value is `3` seconds.
+To detect path failures more quickly, also configure BFD.
+
+|`spec.keepaliveTime`
+|`integer`
+|Optional: Specifies the maximum interval, in seconds, between sending keep-alive messages to the BGP peer.
+If you specify this field, you must also specify a value for the `holdTime` field.
+The specified value must be less than the value for the `holdTime` field.
```
Comment (cgoncalves) [inline networking/metallb/metallb-configure-bgp-peers.adoc:70]: The type of parameters `spec.holdTime` and `spec.keepaliveTime` changed to `string` (converts to `metav1.Duration` in the backend). Examples: `3s`, `1m`, `5m30s`.
Upstream PR: https://github.com/metallb/metallb-operator/pull/132

**Comment by jboxman-rh** [comment]
cc @jboxman-rh

File: modules/nw-metallb-bgppeer-cr.adoc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+|`spec.bfdProfile`
+|`string`
+|Optional: Specifies the name of a BFD profile.
+
+|`spec.nodeSelectors`
+|`object[]`
+|Optional: Specifies a selector, using match expressions and match labels, to control which nodes can connect to the BGP peer.
+
+|`spec.egpMultiHop`
```
Comment (oribon) [inline modules/nw-metallb-bgppeer-cr.adoc:79]: typo: should be **ebgp**MultiHop

File: modules/nw-metallb-bgppeer-cr.adoc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+|`boolean`
+|Optional: Specifies that the BGP peer is multiple network hops away.
```
Comment (oribon) [inline modules/nw-metallb-bgppeer-cr.adoc:81]: I think we should emphasize that it relates only to EBGP peers (when ASN != MyASN)

File: modules/nw-metallb-bgppeer-cr.adoc (hunk 6 above)
Comment (mikemckiernan) [reply modules/nw-metallb-bgppeer-cr.adoc]: Eek.  Fixed and thank you.

File: modules/nw-metallb-bgppeer-cr.adoc (hunk 7 above)
Comment (mikemckiernan) [reply modules/nw-metallb-bgppeer-cr.adoc:81]: Agreed. That was a poor oversight on my part.  I added mention of "different AS" and a one-liner for eBGP.

**Comment by sabinaaledort** [comment]
Just noticed this incorrect part about AdressPools which is in the 4.9 docs as well:
`The list can be a single IP address, a range specified in CIDR notation, a range specified as a starting and ending address separated by a hyphen, or a combination of the three.`
A single IP address is invalid. It resulting an error. For setting a single IP we use a range, for example:
`1.1.1.1-1.1.1.1`

**Comment by skrthomas** [comment]
@asood-rh Could you please QE the PR? In addition to Mike's updates in this PR, in a separate PR I addressed Sabina's comment above with her approval. I also replaced a draft graphic with a graphic from the graphics team. If you have QE feedback in this PR, I'll make the changes in the new PR.

**Pull request closed without merging**